src/Message.ts

```diff
@@ -193,7 +200,48 @@ export default class Message implements proto.MessageV1 {
       msg.error = e
       return msg
     }
-    msg.decrypted = await decrypt(ciphertext, secret, v1Message.headerBytes)
+    msg.decrypted = await decrypt(ciphertext, secret, headerBytes)
     return msg
   }
 }
+
+export class MessageV2 extends Message implements proto.MessageV2 {
+  private header: proto.MessageHeaderV2 // eslint-disable-line camelcase
+  private signed?: proto.SignedContent
+
+  constructor(
+    id: string,
+    bytes: Uint8Array,
+    obj: proto.Message,
+    header: proto.MessageHeaderV2,
+    signed: proto.SignedContent
+  ) {
+    super(id, bytes, obj)
+    this.header = header
+    this.signed = signed
+  }
+
+  static async create(
+    obj: proto.Message,
+    header: proto.MessageHeaderV2,
+    signed: proto.SignedContent,
+    bytes: Uint8Array
+  ): Promise<MessageV2> {
+    const id = bytesToHex(await sha256(bytes))
+    return new MessageV2(id, bytes, obj, header, signed)
+  }
+
+  get sent(): Date {
+    return nsToDate(this.header.createdNs)
+  }
+
+  // wallet address derived from the signature of the message sender
+  get senderAddress(): Promise<string> | undefined {
```

Comment:
MessageV2 doesn't have recipientAddress(), we'll need to figure out how to do without.

Reply:
Seems like we could work it out if we had a reference to the invitation headers. Whichever party was not the signer must be the recipient.

Reply:
It won't work in the context of group conversations, so I was thinking of trying to stop using the concept altogether.

Reply:
Fair enough